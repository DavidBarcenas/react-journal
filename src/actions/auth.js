import { types } from "../types/types";

export const startLoginEmailPwd = ( email, pwd ) => {
  return ( dispatch ) => {
    setTimeout(() => {
      dispatch(login(123, 'Daveepro'))
    }, 3500);    
  }
}

export const login = ( uid, displayName ) => ({
  type: types.login,
  payload: {
    uid, 
    displayName
  }
})