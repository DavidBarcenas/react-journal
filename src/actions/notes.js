import { db } from "../firebase/firebase-provider";
import { types } from "../types/types";
import { loadNotes } from "../helpers/loadNotes";
import Swal from "sweetalert2";
import { fileUpload } from "../helpers/fileUpload";

export const startNewNote = () => {
  return async ( dispatch, getState ) => {
    const {auth: { uid }} = getState();
    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime()
    }
    const doc = await db.collection(`${ uid }/journal/notes`).add( newNote )
    dispatch(activeNote( doc.id, newNote ))
  }
}

export const activeNote = ( id, note ) => ({
  type: types.noteActive,
  payload: {
    id, 
    ...note
  }
})

export const setNotes = ( notes ) => ({
  type: types.noteLoad,
  payload: notes
})

export const startLoadingNotes = ( uid ) => {
  return async ( dispatch ) => {
    const notes = await loadNotes( uid )
    dispatch(setNotes( notes ))
  }
}

export const startSaveNote = ( note ) => {
  return async ( dispatch, getState ) => {
    const {auth: { uid }} = getState();

    if( !note.url ) { delete note.url }

    const noteToFS = { ...note }
    delete noteToFS.id;

    await db.doc(`${ uid }/journal/notes/${ note.id }`).update(noteToFS)
    dispatch(refreshNote( note.id, noteToFS ))
    Swal.fire('Saved', note.title, 'success')
  }
}

export const refreshNote = ( id, note ) => ({
  type: types.noteUpdated,
  payload: {
    id, note: { id, ...note }
  }
})

export const startUploading = ( file ) => {
  return async ( dispatch, getState ) => {
    const { active: activeNote } = getState().notes;
    const fileUrl = await fileUpload( file )

    console.log('fileUrl', fileUrl)
  }
}