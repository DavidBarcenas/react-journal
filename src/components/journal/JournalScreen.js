import React from 'react'
import { Sidebar } from './Sidebar'

export const JournalScreen = () => {
  return (
    <div className="joutnal__wrapper">
      <Sidebar />
      <main><h1>Main content</h1></main>
    </div>
  )
}
