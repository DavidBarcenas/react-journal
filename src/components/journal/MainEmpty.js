import React from 'react'

export const MainEmpty = () => {
  return (
    <div className="main__empty">
      <p>Select something</p>
      <p>or create an entry</p>
      <i className="far fa-star fa-4x"></i>
    </div>
  )
}
