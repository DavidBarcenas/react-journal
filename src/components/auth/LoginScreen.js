import React from 'react'
import { Link } from 'react-router-dom'
import { useDispatch } from 'react-redux'
import { useForm } from '../../hooks/useForm'
import { login } from '../../actions/auth'

export const LoginScreen = () => {

  const dispatch = useDispatch();

  const [ formValues, handleInputChange ] = useForm({
    email: 'davee@gmail.com',
    pwd: '123456'
  });

  const { email, pwd } = formValues;

  const handleLogin = (e) => {
    e.preventDefault();
    dispatch( login(12345, 'Davee') )
  }

  return (
    <>
      <h3 className="auth__title">Login</h3>

      <form onSubmit={ handleLogin }>
        <input type="email" 
          className="auth__input" 
          placeholder="Email" 
          name="email" 
          value={ email } 
          onChange={ handleInputChange }
        />
        <input type="password" 
          className="auth__input" 
          placeholder="Password" 
          name="pwd" 
          value={ pwd } 
          onChange={ handleInputChange } 
        />
        <button type="submit" className="btn btn-primary btn-block">Sign in</button>
      </form>

      <div>
        <span className="auth__social-text">Login with social networks</span>
        <button className="btn btn-block btn__google" >
          <div className="btn__google-img" >
            <img src="/images/google.svg" alt="google"/>
          </div>
          <span>Sign in with google</span>
        </button>

        <Link to="/auth/register" className="auth__social-text"> Create new account </Link>
      </div>
    </>
  )
}
