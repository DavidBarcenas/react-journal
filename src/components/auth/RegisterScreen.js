import React from 'react'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm'
import { useDispatch, useSelector } from 'react-redux';
import isEmail from 'validator/lib/isEmail';
import { setError, removeError } from '../../actions/ui';
import { startRegister } from '../../actions/auth';

export const RegisterScreen = () => {

  const dispatch = useDispatch();
  const { msgError } = useSelector(state => state.ui)

  /* DELETE INITIFORM FOR PRODUCTION ============== */
  const initialForm = {
    name: 'Max',
    email: 'max@mail.com',
    pwd: 'helloworld12',
    pwdConfirm: 'helloworld12',
  }

  const [ formValues, handleInputChange ] = useForm(initialForm);
  const { name, email, pwd, pwdConfirm } = formValues;

  const handleSubmit = (e) => {
    e.preventDefault();
    if( isFormValid() ) {
      dispatch( startRegister( email, pwd, name ) )
    }
  }

  const isFormValid = () => {
    if( name.trim().length === 0 ) {
      dispatch( setError('Name is required') )
      return false;
    } 
    else if( !isEmail( email ) ) {
      dispatch( setError('Email is not valid') )
      return false;
    } 
    else if( pwd !== pwdConfirm || pwd.length < 5 ) {
      dispatch( setError('Password should be at least six characters and match each other') )
      return false;
    }

    dispatch( removeError() )
    return true;
  }

  return (
    <>
      <h3 className="auth__title">Register</h3>

      { msgError && <div className="auth__alert-error"> { msgError} </div> }

      <form onSubmit={ handleSubmit }>
        <input type="text" className="auth__input" placeholder="Name" name="name" value={ name } onChange={ handleInputChange } />
        <input type="email" className="auth__input" placeholder="Email" name="email" value={ email } onChange={ handleInputChange } />
        <input type="password" className="auth__input" placeholder="Password" name="pwd" value={ pwd } onChange={ handleInputChange } />
        <input type="password" className="auth__input" placeholder="Password confirm" name="pwdConfirm" value={ pwdConfirm } onChange={ handleInputChange }/>
        <button type="submit" className="btn btn-primary btn-block">Register</button>
      </form>

      <Link to="/auth/login" className="auth__social-text"> Already registered?</Link>
    </>
  )
}
