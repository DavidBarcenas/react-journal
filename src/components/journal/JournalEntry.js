import React from 'react'
import { useDispatch } from 'react-redux'
import { format } from 'date-fns'
import { activeNote } from '../../actions/notes'

export const JournalEntry = ({ body, date, title, id, url }) => {

  const dispatch = useDispatch()

  const handleEntryClick = () => {
    dispatch(activeNote( id, { body, date, title, id, url } ))
  }
  
  return (
    <div className="animate__animated animate__fadeInUp animate__faster journal__entry" onClick={ handleEntryClick }>
      { url &&
        <div className="journal__entry-picture" style={{backgroundImage: `url(${ url })`}}> </div>
      }
        

      <div className="journal__entry-body">
        <p className="journal__entry-title">{ title }</p>
        <p className="journal__entry-content">{ body }</p>
      </div>

      <div className="journal__entry-date">
        <span className="journal__entry-day">{ format(new Date( date ), "EEEE") }</span>
        <span className="journal__entry-num">{ format(new Date( date ), "io") }</span>
      </div>
    </div>
  )
}
