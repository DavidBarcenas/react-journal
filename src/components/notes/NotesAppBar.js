import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { startSaveNote } from '../../actions/notes'

export const NotesAppBar = () => {

  const dispatch = useDispatch()
  const { active } = useSelector(state => state.notes)

  const handleSave = () => {
    dispatch(startSaveNote( active ))
  }

  return (
    <div className="notes__appbar">
      <span className="notes__appbar-date"> 28 de Agosto de 2020 </span>
      <div>
        <button className="btn"> pic </button>
        <button className="btn" onClick={ handleSave }> Save </button>
      </div>
    </div>
  )
}
