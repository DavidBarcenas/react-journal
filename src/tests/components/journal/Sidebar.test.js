import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import React from 'react'
import { Provider } from 'react-redux';
import { Sidebar } from '../../../components/journal/Sidebar';
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk';
import { startLogout } from '../../../actions/auth';
import { startNewNote } from '../../../actions/notes'

jest.mock('../../../actions/auth', () => ({
  startLogout: jest.fn(),
}))

jest.mock('../../../actions/notes', () => ({
  startNewNote: jest.fn(),
}))

const middlewares = [thunk]
const mockStore = configureStore(middlewares)
const initState = {
  auth: {
    uid: 'd236d',
    name: 'Davee'
  },
  ui: {
    loading: false,
    msgError: null
  },
  notes: {
    notes: [],
    active: null
  }
}
let store = mockStore( initState )
store.dispatch = jest.fn()

const wrapper = mount(
<Provider store={store}>
  <Sidebar />
</Provider>
)

describe('<Sidebar /> testing', () => {
  test('should render correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  test('should trigger startLogout action', () => {
    wrapper.find('.btn').prop('onClick')()
    expect(startLogout).toHaveBeenCalled()
  })
  
  test('should trigger startNewNote action', () => {
    wrapper.find('.journal__new-entry').prop('onClick')()
    expect(startNewNote).toHaveBeenCalled()
  })
  
})
