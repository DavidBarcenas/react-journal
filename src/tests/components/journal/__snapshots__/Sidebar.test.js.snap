// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> testing should render correctly 1`] = `
<aside
  className="journal__sidebar"
>
  <div
    className="journal__sidebar-navbar"
  >
    <h3>
      <i
        className="far fa-moon"
      />
      <span
        className="journal__sidebar-username"
      >
        Davee
      </span>
    </h3>
    <button
      className="btn"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="far fa-calendar-plus fa-5x"
    />
    <p>
      New entry
    </p>
  </div>
  <div
    className="journal__entries"
  />
</aside>
`;
